src/toxcore/dht/server/mod.rs

```diff
@@ -184,25 +173,50 @@ impl Server {
 
         Box::new(pings_stream.for_each(|()| Ok(())))
     }
-    /// send NodesRequest to random peer every 20 seconds
-    pub fn send_nodes_req(&self, friend_pk: PublicKey) -> IoFuture<()> {
-        let mut state = self.state.write();
-        if let Some(peer) = state.kbucket.get_random_node() {
-            let client = state.peers_cache.entry(peer.pk).or_insert_with(ClientData::new);
 
-            let payload = NodesRequestPayload {
-                pk: friend_pk,
-                id: client.new_ping_id(),
-            };
-            let nodes_req = DhtPacket::NodesRequest(NodesRequest::new(
-                &precompute(&peer.pk, &self.sk),
-                &self.pk,
-                payload
-            ));
-            self.send_to(peer.saddr, nodes_req)
-        } else {
-            Box::new(future::ok(()))
+    /// Send NodesRequest to random peer every 20 seconds
+    pub fn periodical_nodes_req(&self) -> IoFuture<()> {
+        let peer = match self.get_random_node() {
+            None => {
+                return Box::new( future::ok(()))
+            },
+            Some(peer) => peer,
+        };
+
+        self.send_nodes_req(peer)
+    }
+    
+    // Get random node from kbucket
+    fn get_random_node(&self) -> Option<PackedNode> {
+        let state = self.state.read();
+        state.kbucket.get_random_node()
+    }
+
+    // Send NodesRequest to peer
+    fn send_nodes_req(&self, target_peer: PackedNode) -> IoFuture<()> {
+        // Check if packet is going to be sent to ourself.
+        if self.pk == target_peer.pk {
+            return Box::new(
+                future::err(
+                    Error::new(ErrorKind::Other, "friend's pk is mine")
+                )
+            )
         }
+
+        let mut state = self.state.write();
+        let client = state.peers_cache.entry(target_peer.pk).or_insert_with(ClientData::new);
+
+        let payload = NodesRequestPayload {
+            pk: target_peer.pk,
```

Comment:
c-toxcore sends request with pk of DHT server. We interested in peers close to us to populate kbucket.

Reply:
In c-toxcore
getnodes be called like this 
        getnodes(dht, assoc_list[rand_node]->ip_port, client_list[rand_node]->public_key, public_key, nullptr);

and getnodes definition is
static int getnodes(DHT *dht, IP_Port ip_port, const uint8_t *public_key, const uint8_t *client_id,
                    const Node_format *sendback_node)


Length  | Content
------- | -------------------------
`1`     | `0x02`
`32`    | Public Key
`24`    | Nonce
`40`    | Payload

first Public Key is DHT server's PK.
and PK in Payload is target peer's PK

Reply:
Nope,
https://github.com/TokTok/c-toxcore/blob/master/toxcore/DHT.c#L1284 - client_id is inserted to the payload,
https://github.com/TokTok/c-toxcore/blob/master/toxcore/DHT.c#L1701 - client_id here is public_key argument,
https://github.com/TokTok/c-toxcore/blob/master/toxcore/DHT.c#L1775 - public_key argument is self DHT public key.

Reply:
https://github.com/TokTok/c-toxcore/blob/7d399cedcfd20f0d91a8caf386ae3c63f4dcf285/toxcore/DHT.c#L1751

Reply:
https://github.com/TokTok/c-toxcore/blob/7d399cedcfd20f0d91a8caf386ae3c63f4dcf285/toxcore/DHT.c#L1774
is for ping with DHT node's PK

Reply:
Now, I found.
NodesRequest is used two ways.
get my own close list.
get my friends's close list.
When get my own close list, it send with my PK.
When get my friend's close list, it send with friend's PK.

Reply:
But you send request to a random peer from kbucket. It's not a friend, so you should look at `do_Close` function instead of `do_DHT_friends`. It's even mentioned in the [spec](https://zetok.github.io/tox-spec/#dht-operation):
> It also sends get node requests to a random node in each of these lists of nodes every 20 seconds, with the search public key being its public key for the closest node and the public key being searched for being the ones in the DHT friends list.

Reply:
Yeah. you are right.